src/devices/sound/discrete.cpp

```diff
@@ -91,61 +91,70 @@ DEFINE_DEVICE_TYPE(DISCRETE, discrete_sound_device, "discrete", "Discrete Sound"
 
 struct output_buffer
 {
+	output_buffer() : source(nullptr), ptr(nullptr)
+	{
+	}
+
+	// required for use in vector (std::atomic<T> has deleted copy constructor)
+	output_buffer(output_buffer &&that) : node_buf(std::move(that.node_buf)), source(that.source), ptr(that.ptr.load(std::memory_order_relaxed)), node_num(that.node_num)
+	{
+	}
+
 	std::unique_ptr<double []>  node_buf;
-	const double                *source;
-	volatile double             *ptr;
+	const double *              source;
+	std::atomic<double *>       ptr;
 	int                         node_num;
```

Comment:
Using `volatile` here doesn’t do what it’s supposed to, as it doesn’t enforce ordering of writes to the buffer relative to writes to the pointer itself.

Making `ptr` atomic guarantees consistent loads on systems where pointer-sized writes are not inherently atomic (typically systems where the memory bus is narrower than a pointer, e.g. 32-bit CPU with 16-bit memory, or 64-bit CPU with 32-bit memory).  It also makes it easier to ensure the compiler doesn’t turn what appears to be a single access int multiple loads.